File: node-api/src/controller/aboutController.js
import * as service from "../services/aboutService.js";
import { Router } from "express";
import { getAuthentication } from "../utils/jwt.js";
import { aboutUpload } from "../config/multer.config.js";
import fs from "fs";
import path from "path";

const endpoints = Router();
const auth = getAuthentication();

endpoints.get("/about", async (req, resp) => {
  const filter = req.query;
  let result = await service.getAbout(filter);
  resp.send(result);
});

endpoints.put(
  "/about/:id",
  auth,
  aboutUpload.single("image"),
  async (req, resp) => {
    try {
      const id = Number(req.params.id);
      const existing = await service.getAboutById(id);

      if (!existing) {
        resp.status(404).send({
          error: "About not found",
        });

        return;
      }

      if (req.file && existing.imageUrl) {
        const oldImagePath = path.resolve(existing.imageUrl.replace("/", ""));

        if (fs.existsSync(oldImagePath)) {
          fs.unlinkSync(oldImagePath);
        }
      }

      const about = {
        presentationText: req.body.presentationText,
        imageUrl: req.file
          ? `/uploads/about/${req.file.filename}`
          : existing.imageUrl,
      };

      const lines = await service.updateAbout(id, about);

      if (lines == 0) {
        resp.status(404).send({
          error: "About not found",
        });

        return;
      }

      resp.send({
        result: "Change saved successfully!",
      });
    } catch (err) {
      resp.status(400).send({
        error: err?.message || "Unexpected error",
      });
    }
  },
);

endpoints.delete("/about/:id", auth, async (req, resp) => {
  let id = Number(req.params.id);
  let lines = await service.deleteAbout(id);

  if (lines == 0) {
    resp.status(404).send({
      error: "About not found",
    });
  } else {
    resp.status(204).send();
  }
});

endpoints.get("/about/:id", auth, async (req, resp) => {
  let id = Number(req.params.id);
  let line = await service.getAboutById(id);

  if (!line) {
    resp.status(404).send({
      error: "About not found",
    });
  } else {
    resp.send(line);
  }
});

export default endpoints;

File: spring-api/projeto-fotografico/projeto-fotografico/src/main/java/com/sp/senac/projeto_fotografico/controllers/AboutController.java
package com.sp.senac.projeto_fotografico.controllers;

import com.sp.senac.projeto_fotografico.models.About;
import com.sp.senac.projeto_fotografico.services.AboutService;
import com.sp.senac.projeto_fotografico.util.FileStorageService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.Map;

@RestController
@RequestMapping("/about")
public class AboutController {

    @Autowired
    private AboutService service;

    @Autowired
    private FileStorageService fileStorage;

    // ─── GET /about — público ────────────────────────────────────────────────

    @GetMapping
    public ResponseEntity<?> getAbout() {
        return ResponseEntity.ok(service.getAbout().orElse(null));
    }

    // ─── GET /about/:id — privado ────────────────────────────────────────────

    @GetMapping("/{id}")
    public ResponseEntity<?> getAboutById(@PathVariable int id) {
        return service.getAboutById(id)
                .<ResponseEntity<?>>map(ResponseEntity::ok)
                .orElse(ResponseEntity.status(404).body(Map.of("error", "About not found")));
    }

    // ─── PUT /about/:id — privado + upload de imagem ─────────────────────────
    //
    // Equivalente ao:  aboutUpload.single("image") + lógica de deleteFile do Node
    // O Angular manda multipart/form-data com os campos:
    //   presentationText → texto
    //   image            → arquivo (opcional)

    @PutMapping(value = "/{id}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResponseEntity<?> update(
            @PathVariable int id,
            @RequestParam("presentationText") String presentationText,
            @RequestParam(value = "image", required = false) MultipartFile image) {

        try {
            About existing = service.getAboutById(id)
                    .orElse(null);

            if (existing == null) {
                return ResponseEntity.status(404).body(Map.of("error", "About not found"));
            }

            String imageUrl = existing.getImageUrl();

            if (image != null && !image.isEmpty()) {
                // Apaga a imagem antiga antes de salvar a nova — igual ao deleteFile() do Node
                if (existing.getImageUrl() != null) {
                    fileStorage.delete(existing.getImageUrl());
                }
                imageUrl = fileStorage.save(image, "about");
            }

            service.update(id, presentationText, imageUrl);
            return ResponseEntity.ok(Map.of("result", "Change saved successfully!"));

        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(404).body(Map.of("error", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    // ─── DELETE /about/:id — privado ─────────────────────────────────────────

    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(@PathVariable int id) {
        try {
            service.delete(id);
            return ResponseEntity.noContent().build();          // 204
        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(404).body(Map.of("error", e.getMessage()));
        }
    }
}
